### ISSUE TITLE
Memory leak when Detect is run multiple times

### ISSUE THREAD

**Issue by HEBOS** [AlexeyAB/darknet#4293]
Hi,

When detect is run multiple times, the memory leak is appearing sporadically, and slowly drains the memory.

My script is modification of the original darknet.py script found in this repository, and can be downloaded [here](https://github.com/HEBOS/darknet/blob/master/darknet.py).

Can you please investigate the issue, as otherwise, I cannot use it in production.

Thank you

**Comment by AlexeyAB** [comment]
Do you get memory leak when you use darknet_video.py?

**Comment by HEBOS** [comment]
I haven't tried darknet_video.py , only darknet.py. Do you want me to try that one? 
What would make the difference?

**Comment by AlexeyAB** [comment]
Yes. To localize the error.

**Comment by HEBOS** [comment]
I have tested darknet_video.py too, and here are the results.

running 20 sec = loss 651264
running 30 sec = loss 667648
running 100 sec = we have gain i.e. loss = -25370624, which leads me to conclusion that it might be that the wrong locations are being deallocated.

Here is my [script ](https://github.com/HEBOS/darknet/blob/master/darknet_video.py), and the [commit](https://github.com/HEBOS/darknet/commit/a0d09b76d4299917101463aecde5fe781eadc6be), so you can easily see, what I have added there.

Many thanks

**Comment by AlexeyAB** [comment]
> loss = -25370624

Does it mean that there is a memory leak in darknet_video.py - each 100 seconds 25 MB ?

Do you use Windows or Linux?

What params do you use in the Makefile?

**Comment by HEBOS** [comment]
No.
If we run the test only 20 sec, we have memory leak 651264 bytes.
If we run it 30 seconds, we have memory leak 667648 bytes.
However, if we run it 100 seconds, we have memory gain 25 MB, which is an absurd.

**Comment by HEBOS** [comment]
NAME="Ubuntu"
VERSION="18.04.3 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.3 LTS"
VERSION_ID="18.04"

I have only amended destination architecture to include Jetson Nano.
You can see my [commit](https://github.com/HEBOS/darknet/commit/755438fe165e8e7bf85b7f655089359c90afd847).

The results are completely unpredictable. When I run 100 seconds test for the second time, I'm getting almost 1MB memory leak (1007616 bytes).

**Comment by AlexeyAB** [comment]
What params do you use in the Makefile?
Do you use GPU=1 CUDNN=1 OPENCV=1 DEBUG=0 or something else?

**Comment by HEBOS** [comment]
The same as your original file.

GPU=1
CUDNN=1
CUDNN_HALF=0
OPENCV=1
AVX=0
OPENMP=0
LIBSO=1
ZED_CAMERA=0
DEBUG=0

**Comment by AlexeyAB** [comment]
> However, if we run it 100 seconds, we have memory gain 25 MB, which is an absurd.

* So there is no memory leak in darknet_video.py?

* Try to compile darknet library with OPENCV=0

* What size of memory leak do you get after several hours by using your own python script?

* How many memory your script + darknet-library takes after several hours - look by using `top` command?

* How long can an application run until all memory runs out?

Some memory can be spend by OpenCV (ffmpeg, ...) or OS Linux.

**Comment by LukeAI** [comment]
@HEBOS  which commit are you on? There is definitely a recent memory leak in a recent commit. :/

**Comment by LukeAI** [comment]
OK, the leak does not exist in 3652d7d3745a29ab161cf3a77492f1ac8b04b090

**Comment by LukeAI** [comment]
The leak does not exist in c516b6cb0a08f82023067a649d10238ff18cf1e1

**Comment by LukeAI** [comment]
ok, so certainly the memory leak was introduced in commit 11142d00bedbafb015991fb20a05a5eb048200d6

**Comment by LukeAI** [comment]
I didn't notice it earlier because I was training very small networks and it didn't saturate my 32GB until training was finished. When I tried training slightly larger networks, they crashed before training completion.

**Comment by AlexeyAB** [comment]
> ok, so certainly the memory leak was introduced in commit 11142d0

I just fixed it: https://github.com/AlexeyAB/darknet/commit/509ba13acf9b6738d9838eb1553a2f40e36c0c2e

**Comment by LukeAI** [comment]
you respond incredibly quickly to bug reports.

**Comment by LukeAI** [comment]
OK, I think that the classifier training memory leak was actually unrelated to OPs memory leak as I am seeing the same behaviour.

**Comment by AlexeyAB** [comment]
> -> Error is still present in darknet_video.py, but the results are unpredictable.

What does it mean?


> If I turn detection on, it starts leaking, and in 30 seconds, the memory is full, and the app crashes.

How many seconds can the **default** darknet_video.py work before it crashes?

**Comment by AlexeyAB** [comment]
@HEBOS Also look at the Valgrind output about memory leaks: https://github.com/AlexeyAB/darknet/issues/3633#issuecomment-554356260

**Comment by HEBOS** [comment]
Unpredictable result means that the results are not consistent, when I repeat the test.
I have started 40 mins test (that's the length of the video), and will let you know the results.

In the mean time, can you please check the results of cppcheck tool:

[src/classifier.c:1291]: (error) Memory leak: gpus
	[src/coco.c:230]: (error) Memory leak: val
	[src/coco.c:230]: (error) Memory leak: val_resized
	[src/coco.c:230]: (error) Memory leak: thr
	[src/coco.c:314]: (error) Memory leak: fps
	[src/detector.c:569]: (error) Memory leak: fps
	[src/detector.c:569]: (error) Memory leak: val
	[src/detector.c:569]: (error) Memory leak: val_resized
	[src/detector.c:569]: (error) Memory leak: thr
	[src/detector.c:1103]: (error) Memory leak: val
	[src/detector.c:1103]: (error) Memory leak: val_resized
	[src/detector.c:1103]: (error) Memory leak: thr
	[src/detector.c:1508]: (error) Memory leak: gpus
	[src/image_opencv.cpp:722]: (error) Memory leak: src
	[src/network.c:839]: (error) Memory leak: buf
	[src/utils.c:690]: (error) Memory leak: values
	[src/utils.c:690]: (error) Memory leak: indexes
	[src/utils.c:68]: (error) Memory leak: swp
	[src/yolo.c:193]: (error) Memory leak: fps
	[src/yolo.c:193]: (error) Memory leak: val
	[src/yolo.c:193]: (error) Memory leak: val_resized
	[src/yolo.c:193]: (error) Memory leak: thr
	[src/yolo.c:276]: (error) Memory leak: fps
	[src/yolo_v2_class.cpp:177]: (error) Memory leak: detector_gpu.avg
	[src/yolo_v2_class.cpp:177]: (error) Memory leak: detector_gpu.track_id

	[src/data.c:204]: (error) Common realloc mistake: 'boxes' nulled but not freed upon failure
	[src/detector.c:829]: (error) Common realloc mistake: 'detections' nulled but not freed upon failure
	[src/detector.c:1177]: (error) Common realloc mistake: 'rel_width_height_array' nulled but not freed upon failure
	[src/network.c:838]: (error) Common realloc mistake: 'send_buf' nulled but not freed upon failure
	[src/rnn.c:24]: (error) Common realloc mistake: 'd' nulled but not freed upon failure
	[src/rnn.c:30]: (error) Common realloc mistake: 'd' nulled but not freed upon failure
	[src/rnn.c:46]: (error) Common realloc mistake: 'd' nulled but not freed upon failure
	[src/rnn.c:51]: (error) Common realloc mistake: 'd' nulled but not freed upon failure
	[src/utils.c:41]: (error) Common realloc mistake: 'map' nulled but not freed upon failure


	[src/go.c:50]: (error) Resource leak: fp
	[src/utils.c:44]: (error) Resource leak: file

	[src/gemm.c:327]: (error) Shifting signed 32-bit value by 31 bits is undefined behaviour
	[src/gemm.c:668]: (error) Shifting signed 32-bit value by 31 bits is undefined behaviour

**Comment by AlexeyAB** [comment]
@HEBOS I fixed some memory leak bugs from your list: https://github.com/AlexeyAB/darknet/commit/3555beb91401cf746fd357b4029231143d00f58f

But it doesn't related to `./darknet detector test / demo / train ... `
except:
`gpus` - you can lost maximum several bytes

So if you used `./darknet detector test / demo / train ... ` then there were no memory leaks more than 100 bytes in total.

----

`[src/image_opencv.cpp:722]: (error) Memory leak: src` - is related to the `./darknet classifier demo ...` - This is really a bug.


`[src/utils.c:44]: (error) Resource leak: file` - is a minor bug for Yolo9000

`[src/utils.c:68]: (error) Memory leak: swp` - is a bug for `./darknet compare ...`


`[src/utils.c:690]: (error) Memory leak: values [src/utils.c:690]: (error) Memory leak: indexes` -  is a minor bug for `./darknet rnn ...` but currently it was commented

```
[src/detector.c:569]: (error) Memory leak: fps
[src/detector.c:569]: (error) Memory leak: val
[src/detector.c:569]: (error) Memory leak: val_resized
[src/detector.c:569]: (error) Memory leak: thr
[src/detector.c:1103]: (error) Memory leak: val
[src/detector.c:1103]: (error) Memory leak: val_resized
[src/detector.c:1103]: (error) Memory leak: thr
[src/detector.c:1508]: (error) Memory leak: gpus
```
These are for `./darknet detector valid / map ...`

----

```
[src/yolo_v2_class.cpp:177]: (error) Memory leak: detector_gpu.avg
[src/yolo_v2_class.cpp:177]: (error) Memory leak: detector_gpu.track_id
```
There is no memory leak: https://github.com/AlexeyAB/darknet/blob/3555beb91401cf746fd357b4029231143d00f58f/src/yolo_v2_class.cpp#L185-L187

----
`[src/network.c:839]: (error) Memory leak: buf`
and
```
[src/data.c:204]: (error) Common realloc mistake: 'boxes' nulled but not freed upon failure
[src/detector.c:829]: (error) Common realloc mistake: 'detections' nulled but not freed upon failure
[src/detector.c:1177]: (error) Common realloc mistake: 'rel_width_height_array' nulled but not freed upon failure
[src/network.c:838]: (error) Common realloc mistake: 'send_buf' nulled but not freed upon failure
[src/rnn.c:24]: (error) Common realloc mistake: 'd' nulled but not freed upon failure
[src/rnn.c:30]: (error) Common realloc mistake: 'd' nulled but not freed upon failure
[src/rnn.c:46]: (error) Common realloc mistake: 'd' nulled but not freed upon failure
[src/rnn.c:51]: (error) Common realloc mistake: 'd' nulled but not freed upon failure
[src/utils.c:41]: (error) Common realloc mistake: 'map' nulled but not freed upon failure
```
But memory leak with such `Common realloc mistake` will be only after all virtual memory is depleted, when any program should terminate - so it will not matter if there is a leak or not

**Comment by HEBOS** [comment]
I grabbed your commit, and the situation is even worse.
Memory leak, after 40 seconds is around 44 MB.
You can get my updated detector_video.py script from [here](https://github.com/HEBOS/darknet/blob/master/darknet_video.py).

Here is the log:
	Memory free before 2 detection: 1295544320
	Memory free after 2 detection: 1294053376, loss 1490944

	Memory free before 3 detection: 1290117120
	Memory free after 3 detection: 1289863168, loss 253952

	Memory free before 4 detection: 1289355264
	Memory free after 4 detection: 1289084928, loss 270336

	Memory free before 5 detection: 1287688192
	Memory free after 5 detection: 1287274496, loss 413696

	Memory free before 6 detection: 1286766592
	Memory free after 6 detection: 1285971968, loss 794624

	Memory free before 7 detection: 1286037504
	Memory free after 7 detection: 1286672384, loss -634880

	Memory free before 8 detection: 1286672384
	Memory free after 8 detection: 1283670016, loss 3002368

	Memory free before 9 detection: 1285783552
	Memory free after 9 detection: 1286164480, loss -380928

	Memory free before 10 detection: 1286164480
	Memory free after 10 detection: 1286258688, loss -94208

	Memory free before 11 detection: 1286258688
	Memory free after 11 detection: 1286275072, loss -16384

	Memory free before 12 detection: 1286275072
	Memory free after 12 detection: 1286529024, loss -253952

	Memory free before 13 detection: 1286529024
	Memory free after 13 detection: 1286656000, loss -126976

	Memory free before 14 detection: 1286275072
	Memory free after 14 detection: 1286098944, loss 176128

	Memory free before 15 detection: 1285971968
	Memory free after 15 detection: 1285767168, loss 204800

	Memory free before 16 detection: 1285767168
	Memory free after 16 detection: 1285386240, loss 380928

	Memory free before 17 detection: 1285259264
	Memory free after 17 detection: 1285226496, loss 32768

	Memory free before 18 detection: 1284829184
	Memory free after 18 detection: 1283780608, loss 1048576

	Memory free before 19 detection: 1284161536
	Memory free after 19 detection: 1284956160, loss -794624

	Memory free before 20 detection: 1284956160
	Memory free after 20 detection: 1284591616, loss 364544

	Memory free before 21 detection: 1284591616
	Memory free after 21 detection: 1284972544, loss -380928

	Memory free before 22 detection: 1284972544
	Memory free after 22 detection: 1285066752, loss -94208

	Memory free before 23 detection: 1285066752
	Memory free after 23 detection: 1284669440, loss 397312

	Memory free before 24 detection: 1284669440
	Memory free after 24 detection: 1284067328, loss 602112

	Memory free before 25 detection: 1284067328
	Memory free after 25 detection: 1284448256, loss -380928

	Memory free before 26 detection: 1284448256
	Memory free after 26 detection: 1284812800, loss -364544

	Memory free before 27 detection: 1284812800
	Memory free after 27 detection: 1284431872, loss 380928

	Memory free before 28 detection: 1284431872
	Memory free after 28 detection: 1284177920, loss 253952

	Memory free before 29 detection: 1284177920
	Memory free after 29 detection: 1283923968, loss 253952

	Memory free before 30 detection: 1283923968
	Memory free after 30 detection: 1283383296, loss 540672

	Memory free before 31 detection: 1283383296
	Memory free after 31 detection: 1283162112, loss 221184

	Memory free before 32 detection: 1282891776
	Memory free after 32 detection: 1283670016, loss -778240

	Memory free before 33 detection: 1283670016
	Memory free after 33 detection: 1283620864, loss 49152

	Memory free before 34 detection: 1283620864
	Memory free after 34 detection: 1283256320, loss 364544

	Memory free before 35 detection: 1283272704
	Memory free after 35 detection: 1283510272, loss -237568

	Memory free before 36 detection: 1283510272
	Memory free after 36 detection: 1283129344, loss 380928

	Memory free before 37 detection: 1283145728
	Memory free after 37 detection: 1283493888, loss -348160

	Memory free before 38 detection: 1282969600
	Memory free after 38 detection: 1283891200, loss -794624

	Memory free before 39 detection: 1283764224
	Memory free after 39 detection: 1284145152, loss -380928

	Memory free before 40 detection: 1284288512
	Memory free after 40 detection: 1284255744, loss 32768

	Memory free before 41 detection: 1284255744
	Memory free after 41 detection: 1284653056, loss -397312

	Memory free before 42 detection: 1284653056
	Memory free after 42 detection: 1284239360, loss 413696

	Memory free before 43 detection: 1284239360
	Memory free after 43 detection: 1284747264, loss -507904

	Memory free before 44 detection: 1284747264
	Memory free after 44 detection: 1284349952, loss 397312

	Memory free before 45 detection: 1284349952
	Memory free after 45 detection: 1284464640, loss -114688

	Memory free before 46 detection: 1284464640
	Memory free after 46 detection: 1284517888, loss -53248

	Memory free before 47 detection: 1284517888
	Memory free after 47 detection: 1284468736, loss 49152

	Memory free before 48 detection: 1284468736
	Memory free after 48 detection: 1284263936, loss 204800

	Memory free before 49 detection: 1284263936
	Memory free after 49 detection: 1284390912, loss -126976

	Memory free before 50 detection: 1284390912
	Memory free after 50 detection: 1284136960, loss 253952

	Memory free before 51 detection: 1284136960
	Memory free after 51 detection: 1284501504, loss -364544

	Memory free before 52 detection: 1284501504
	Memory free after 52 detection: 1284120576, loss 380928

	Memory free before 53 detection: 1284120576
	Memory free after 53 detection: 1284485120, loss -364544

	Memory free before 54 detection: 1284485120
	Memory free after 54 detection: 1283739648, loss 745472

	Memory free before 55 detection: 1283866624
	Memory free after 55 detection: 1284136960, loss -270336

	Memory free before 56 detection: 1284136960
	Memory free after 56 detection: 1284247552, loss -110592

	Memory free before 57 detection: 1284247552
	Memory free after 57 detection: 1284628480, loss -380928

	Memory free before 58 detection: 1284628480
	Memory free after 58 detection: 1283944448, loss 684032

	Memory free before 59 detection: 1283817472
	Memory free after 59 detection: 1284501504, loss -684032

	Memory free before 60 detection: 1284771840
	Memory free after 60 detection: 1284898816, loss -126976

	Memory free before 61 detection: 1284898816
	Memory free after 61 detection: 1284374528, loss 524288

	Memory free before 62 detection: 1284374528
	Memory free after 62 detection: 1284644864, loss -270336

	Memory free before 63 detection: 1284644864
	Memory free after 63 detection: 1284771840, loss -126976

	Memory free before 64 detection: 1284771840
	Memory free after 64 detection: 1284374528, loss 397312

	Memory free before 65 detection: 1284374528
	Memory free after 65 detection: 1284104192, loss 270336

	Memory free before 66 detection: 1284104192
	Memory free after 66 detection: 1284898816, loss -794624

	Memory free before 67 detection: 1284898816
	Memory free after 67 detection: 1284771840, loss 126976

	Memory free before 68 detection: 1284739072
	Memory free after 68 detection: 1284771840, loss -32768

	Memory free before 69 detection: 1284771840
	Memory free after 69 detection: 1284771840, loss 0

	Memory free before 70 detection: 1284628480
	Memory free after 70 detection: 1284898816, loss -270336

	Memory free before 71 detection: 1284898816
	Memory free after 71 detection: 1284612096, loss 286720

	** Memory leak after all detections: 34873344 **

**Comment by AlexeyAB** [comment]
* How many seconds can the default darknet_video.py work before it crashes?

* What size of memory leak do you get after several hours by using your own python script?

**Comment by HEBOS** [comment]
I'm going to close the issue, as I can confirm that there is no more memory leak.
After running darknet_video.py for 40 minutes, the amount of free memory remained the same.
